master/buildbot/process/buildstep.py

```diff
@@ -789,6 +789,8 @@ def getWorkerName(self):
         return self.build.getWorkerName()
 
     def addLog(self, name, type='s', logEncoding=None):
+        if self.stepid is None:
```

Comment:
looks like unrelated

Reply:
This is the root cause of the problem. 

Since stepid is None until it is actually added to the db, if a build is cancelled before the build step is properly initialized, there may be a call to addLog while stepid is still None.

Now that there are NOT NULL constraints on the logs.stepid column, we should not experience db integrity problems. However, this would cause errors to be risen.

Reply:
Please note that I took care of putting this in a separate commit.

Reply:
A better PR description would have prevented this misunderstanding :-)

Reply:
You're being polite :-)

```bash
| sed 's/better //'
```

PR description updated.

Reply:
@tardyp: Could you re-review the PR? Then I'll merge it as soon as we have tests.